### ISSUE TITLE
32X emulation - emscripten

### ISSUE THREAD

**Issue by ethanaobrien** [libretro/picodrive#233]
Hello all,

I've been working on trying to get 32X emulation working in the browser, however it appears to just load to a black screen (specifically 32x emulation). I've traced this back to a core issue, and after hours of digging through the program, I've came here for help.

The portion of the logs that caught my eye was particularly this section:
![image](https://github.com/user-attachments/assets/cbe4ecc6-f1e0-4b3d-8dd4-a47f238ee1a2)

As this doesnt appear to occur on a native copy of RetroArch I've compiled.

Changes I've made so far:
[Told libretro to not use mmap](https://github.com/libretro/picodrive/compare/master...EmulatorJS:picodrive:master#diff-59740492c99816931aaa08996521d33e4c0ca52b1186c1074f465658de460152), since mmap location hints are not supported in emscripten.
My current `emscripten` target options are:
```
        TARGET := $(TARGET_NAME)_libretro_$(platform).bc
        ARCH = unknown

        STATIC_LINKING = 1
        STATIC_LINKING_LINK = 1
```

Any information leading to where I can get more logs, and what I can do to fix the issue would be appreciated.
Thank you

Edit: as a note, this is somehow possible, as there are several closed source versions of picodrive with working emscripten support ([emulatorjs.com](https://www.emulatorjs.com/core/sega32x.html)). Through some code injection I have managed to get the RetroArch/picodrive logs, and no `ptr is not aliqned` errors were present

**Comment by irixxxx** [comment]
I think the alignment for code pointers isn't working correctly. The Makefile sets ``-falign-functions=2`` since the lowest bit of every code pointer must be 0. That bit is used to discriminate code pointers from memory pointers in a fast way.
Check that the ``-falign-functions`` option is set during compilation, or check for warnings from the compiler that it is ignored.

**Comment by ethanaobrien** [comment]
Got it, I've fixed that now.

New issue, I get a `null function or function signature mismatch` after running for a seconds which I traced to here https://github.com/libretro/picodrive/blob/479f120b928153815a4543b7c4104c0aa355a660/cpu/cz80/cz80.c#L285

How should I debug this?

**Comment by irixxxx** [comment]
The easiest i can think off is to replace the include line with the whole content of the referenced file. That should give you a proper line number in the crash.

**Comment by irixxxx** [comment]
Closing as there has been no feedback for month. Besides, there's an emscripten version of picodrive build automatically by buildbot and that compiles just fine.

**Comment by ethanaobrien** [comment]
It compiles just fine, but Sega 32x is completely broken. I haven't had time to look into this due to being busy, though I hope to get around to it

**Comment by irixxxx** [comment]
If you have now info, feel free to comment, reopen, or make a PR, please!

**Comment by ethanaobrien** [comment]
Finally got time to sit down and take a look @irixxxx 

It turns out these lines are causing a segfault on emscripten. Not sure why, since I'm not familiar enough with the project: https://github.com/libretro/picodrive/blob/09d3cb634c337aa3a1ed6575267d37bc1fdf064e/cpu/cz80/cz80.c#L223-L224

However adding a pre-processor to not run that code on emscripten seems to fix it entirely? I'm not sure if this has any unwanted side-affects, though

Emscripten doesn't (yet) support function alignment, though I'm working on getting it upstreamed to wasm-ld.

**Comment by irixxxx** [comment]
Picodrive relies on the lowest bit being zero in the set of memory access functions, to be able to distinguish a direct mapping through a memory pointer from an access function. That's so ingrained in it that it's probably way too much work to change this. 

So, it'll probably be best to upstream your changes to emscripten first.

**Comment by ethanaobrien** [comment]
I'm in the process of doing so, waiting on help for this PR: https://github.com/llvm/llvm-project/pull/105532

Once that PR is merged, it won't require any change in this repo, because functions need to be aligned at link time and emscripten can't (currently) ship as a shared object library
